compiler/src/iree/compiler/Codegen/Common/MaterializeTuningSpecsPass.cpp

```diff
@@ -101,37 +164,69 @@ struct MaterializeTuningSpecsPass final
-        dialect->getOrLoadTransformLibraryModule(clCodegenTuningSpecPath);
-    if (failed(maybeTransformLibrary)) {
-      module->emitError()
-          << "Failed to load tuning spec transform dialect library from "
-          << clCodegenTuningSpecPath;
+    assert(dialect);
+
+    FailureOr<ModuleOp> userTuningSpec = getUserTuningSpec(module, *dialect);
+    const bool hasUserTuningSpec = succeeded(userTuningSpec);
+    if (!hasUserTuningSpec && !clCodegenTuningSpecPath.empty()) {
+      // When a user spec is requested but fails to load, this is a hard
+      // failure.
       return signalPassFailure();
     }
 
-    ModuleOp userTuningSpec = *maybeTransformLibrary;
-    if (!userTuningSpec.getSymName()) {
-      // Set a module name so that we can refer to its nested symbols.
-      userTuningSpec.setSymName("iree_user_tuning_spec");
+    FailureOr<ModuleOp> defaultTuningSpec =
+        getDefaultTuningSpec(module, *dialect);
+    const bool hasDefaultTuningSpec = succeeded(defaultTuningSpec);
+    if (!hasUserTuningSpec && !hasDefaultTuningSpec) {
+      // No specs available, nothing to do.
+      return;
     }
 
-    Location loc = userTuningSpec.getLoc();
+    // If only the default tuning spec is available, use it directly and skip
+    // the linking stage.
+    if (!hasUserTuningSpec) {
+      if (failed(dumpFinalTuningSpecToDir(*defaultTuningSpec))) {
+        return signalPassFailure();
+      }
+      FailureOr<DenseElementsAttr> serializedSpec =
+          serializeTuningSpecToAttr(*defaultTuningSpec);
+      if (failed(serializedSpec)) {
+        module->emitError("Failed to serialize default tuning specs");
+        return signalPassFailure();
+      }
+      module->setAttr(kSerializedTuningSpecAttrName, *serializedSpec);
+      return;
+    }
+
+    // When the user tuning spec is available, link all available libraries into
+    // a single module. We insert the default tuning spec last, so that any
+    // user-specified tuning configurations take precedence.
+    SmallVector<ModuleOp, 2> allSpecs = {*userTuningSpec};
+    if (hasDefaultTuningSpec) {
+      allSpecs.push_back(*defaultTuningSpec);
+    }
+
+    Location loc =
+        FusedLoc::get(ctx, llvm::map_to_vector<2>(allSpecs, [](ModuleOp m) {
+                        return m.getLoc();
+                      }));
 
     // This module will always be released at the end of the pass.
     OwningOpRef<ModuleOp> linkedTuningSpec(
         ModuleOp::create(loc, "iree_linked_tuning_spec"));
     linkedTuningSpec.get()->setAttr(
         transform::TransformDialect::kWithNamedSequenceAttrName,
         UnitAttr::get(ctx));
-    linkedTuningSpec->insert(linkedTuningSpec->begin(), userTuningSpec.clone());
+    for (auto [idx, spec] : llvm::enumerate(allSpecs)) {
+      ModuleOp clonedSpec = spec.clone();
+      // Make sure there are no symbol name collisions.
+      clonedSpec.setSymName(
+          llvm::formatv("{}_{}", clonedSpec.getSymName().value(), idx).str());
```

Comment:
Neat - I didn't realize that `llvm::formatv` allows standard `{}` as opposed to the numbered placeholders like `{0}`, `{1}` that it suggests in its own comments. Should do a pass to align on that -- makes `llvm::formatv` drop-in replaceable by `std::format` / `fmt::format` whenever we need to.

Reply:
Actually I just realized that `{0}`, `{1}` is also standard. TIL!

Reply:
https://github.com/iree-org/iree/pull/19426